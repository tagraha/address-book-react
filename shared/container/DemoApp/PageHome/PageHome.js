import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withJob } from 'react-jobs';
import debounce from 'lodash.debounce';

import { loadInitialUsers, loadUsers } from './../../../redux/modules/users';
import UsersCard from '../../../components/UsersCard/UsersCard';

import config from '../../../../config';

class PageHome extends React.Component {
  constructor(props) {
    super(props);
    this.handleOnScroll = this.handleOnScroll.bind(this);
  }
  componentDidMount() {
    window.addEventListener('scroll', this.handleOnScroll);
  }
  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleOnScroll);
  }
  handleOnScroll() {
    var scrollTop = (document.documentElement && document.documentElement.scrollTop) || document.body.scrollTop;
    var scrollHeight = (document.documentElement && document.documentElement.scrollHeight) || document.body.scrollHeight;
    var clientHeight = document.documentElement.clientHeight || window.innerHeight;
    var scrolledToBottom = Math.ceil(scrollTop + clientHeight) >= scrollHeight;

    if (scrolledToBottom && !this.props.isLoading) {
      this.props.loadUsers();
    }
  }
  render() {
    const { users } = this.props;
    return (
      <React.Fragment>
        <Helmet>
          <title>Home</title>
        </Helmet>

        <div className="row center-xs">
          <div className="col-xs-6">
            <div className="box">

              <div className="row start-xs">
                <div className="col-xs-12">
                  <div className="box">
                    <h2>{config('welcomeMessage')}</h2>
                    <Link to="/setting">setting</Link>
                    {users && users.map((value) => (
                      <UsersCard user={value} key={value.email} />
                    ))}
                  </div>
                </div>
              </div>

            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => ({
  isError: state.users.data.isError,
  isLoading: state.users.data.isLoading,
  users: state.users.data.results,
});

const mapActionsToProps = {
  loadInitialUsers: loadInitialUsers,
  loadUsers: loadUsers
};

PageHome.propTypes = {};

export default compose(
  connect(mapStateToProps, mapActionsToProps),
  withJob({
    work: (
      { match, users, loadInitialUsers }, // eslint-disable-line
    ) => {
      // naive caching
      if (users.length > 0) {
        return false
      }
      // Execute the redux-thunk powered action that returns a Promise and
      // fetches the users.
      loadInitialUsers()
      // dont need to trigger the work. unless you need re-render the page with new data
      // ex: navigating from (current page) /article/news-1 to /article/news-2
      // we need to update with news-2 data on the same page/component right? then we need
      // to trigger the work and pass news-2 parameter; update redux state tree.
      shouldWorkAgain: (prevProps, nextProps) => {
        // eslint-disable-line
        return false;
      }
    }
  }),
)(PageHome);
