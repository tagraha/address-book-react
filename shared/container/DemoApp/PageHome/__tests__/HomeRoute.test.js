/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';

import PageHome from '../PageHome';

describe('should render <PageHome /> correctly', () => {
  test('renders', () => {
    const wrapper = shallow(<PageHome />);
    expect(wrapper).toMatchSnapshot();
  });
});
