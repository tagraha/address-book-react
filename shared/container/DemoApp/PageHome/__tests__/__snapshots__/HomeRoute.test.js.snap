// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <PageHome /> correctly renders 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Home
    </title>
  </HelmetWrapper>
  <h2>
    Hello world!
  </h2>
  <p>
    this is the homepage
  </p>
</div>
`;
